File: src/python/lexer.py
from collections import namedtuple
from dataclasses import dataclass
from enum import Enum
import logging


TokenType = Enum("TokenType", [
    "INTEGER", "FLOAT", "RATIONAL", "STRING", "CHAR", "BOOLEAN", "IDENTIFIER",
    "EQUALS_EQUALS", "BANG_EQUALS",
    "LESS_THAN", "GREATER_THAN", "LESS_EQUALS", "GREATER_EQUALS",
    "EQUALS", "STAR", "SLASH", "PLUS", "MINUS", "BANG",
    "SEMICOLON", "COLON", "COMMA", "DOT",
    "LEFT_BRACKET", "RIGHT_BRACKET",
    "LEFT_CURLY", "RIGHT_CURLY", "LEFT_SQUARE", "RIGHT_SQUARE",
    "IF", "ELSE", "WHILE", "RETURN",
    "VAR", "CONST", "FUNC", "STRUCT", "INTERFACE", "IMPL",
    "ARR", "VEC",
])
Token = namedtuple("Token", ["typ", "lexeme", "pos"])


@dataclass
class Pos:
    line: int
    col: int

    def __iadd__(self, other):
        return Pos(self.line, self.col + other)


T = TokenType


KEYWORDS = {
    "true": T.BOOLEAN,
    "false": T.BOOLEAN,
    "if": T.IF,
    "else": T.ELSE,
    "while": T.WHILE,
    "return": T.RETURN,
    "var": T.VAR,
    "const": T.CONST,
    "fn": T.FUNC,
    "struct": T.STRUCT,
    "interface": T.INTERFACE,
    "impl": T.IMPL,
    "arr": T.ARR,
    "vec": T.VEC,
}

DIGRAPHS = {
    "==": T.EQUALS_EQUALS,
    "!=": T.BANG_EQUALS,
    "<=": T.LESS_EQUALS,
    ">=": T.GREATER_EQUALS,
}
MONOGRAPHS = {
    "<": T.LESS_THAN,
    ">": T.GREATER_THAN,
    "=": T.EQUALS,
    "*": T.STAR,
    "/": T.SLASH,
    "+": T.PLUS,
    "-": T.MINUS,
    "!": T.BANG,
    ";": T.SEMICOLON,
    ":": T.COLON,
    ",": T.COMMA,
    ".": T.DOT,
    "(": T.LEFT_BRACKET,
    ")": T.RIGHT_BRACKET,
    "{": T.LEFT_CURLY,
    "}": T.RIGHT_CURLY,
    "[": T.LEFT_SQUARE,
    "]": T.RIGHT_SQUARE,
}

SEMICOLON_INSERT = [
    T.INTEGER, T.FLOAT, T.RATIONAL, T.STRING, T.CHAR, T.BOOLEAN, T.IDENTIFIER,
    T.RIGHT_BRACKET, T.RIGHT_SQUARE,
    T.RETURN,
]


def get(src, index, length=1):
    if index + length <= len(src):
        return src[index:index+length]
    else:
        return ""


def error(errh, msg):
    if errh is None:
        assert False, msg
    else:
        errh(msg)


def tokenise(src, error_handler=None):
    cur = 0
    pos = Pos(1, 1)
    tokens = []
    while cur < len(src):
        cur_pos = pos
        lexeme = None
        typ = None
        char = src[cur]
        if char == '\n':
            pos = Pos(pos.line + 1, 1)
            # automatic semicolon insertion
            if tokens and tokens[-1].typ in SEMICOLON_INSERT:
                lexeme = ';'
                typ = T.SEMICOLON
            else:
                cur += 1
                continue
        elif char.isspace():
            pos += 1
            cur += 1
            continue
        elif char.isalpha() or char == '_':
            # identifier or keyword (including bool or type)
            i = 0
            while (c := get(src, cur + i)).isalnum() or c == '_':
                i += 1
            lexeme = src[cur:cur + i]
            typ = KEYWORDS.get(lexeme) or T.IDENTIFIER
        elif char.isnumeric():
            # int or float
            i = 0
            typ = T.INTEGER
            while get(src, cur + i).isnumeric():
                i += 1
                if typ == T.INTEGER:
                    if (
                        get(src, cur + i) == '.' and
                        get(src, cur + i + 1).isnumeric()
                    ):
                        typ = T.FLOAT
                        i += 1
                    elif (
                        get(src, cur + i, 2) == '//' and
                        get(src, cur + i + 2).isnumeric()
                    ):
                        typ = T.RATIONAL
                        i += 2
            lexeme = src[cur:cur + i]
        elif char == '"':
            # string
            i = 1
            while get(src, cur + i) != '"':
                i += 1
                if i > len(src):
                    msg = "Syntax error: unterminated string literal"
                    error(error_handler, msg)
            i += 1
            lexeme = src[cur:cur + i]
            typ = T.STRING
        elif char == "'":
            # character
            logging.debug(src[cur:cur + 2])
            if get(src, cur + 2) != "'":
                msg = "Syntax error: invalid char literal"
                error(error_handler, msg)
            lexeme = src[cur:cur + 3]
            typ = T.CHAR
        else:
            # all the easy tokens
            for monograph, token in MONOGRAPHS.items():
                if char == monograph:
                    lexeme = monograph
                    typ = token
                    break
            for digraph, token in DIGRAPHS.items():
                if char + get(src, cur + 1) == digraph:
                    lexeme = digraph
                    typ = token
                    break

        if not lexeme:
            # syntax error (unexpected char)
            msg = f"Syntax error: unexpected character '{char}'"
            error(error_handler, msg)
        tokens.append(Token(typ, lexeme, cur_pos))
        pos += len(lexeme)
        cur += len(lexeme)
    return tokens


if __name__ == "__main__":
    import sys
    if len(sys.argv) == 2:
        src_file = sys.argv[1]
    else:
        src_file = "input.txt"
    with open(src_file) as f:
        src = f.read()
    tokens = tokenise(src)
    print(tokens)

File: src/zig/lexer.zig
const std = @import("std");
const print = std.debug.print;

pub const Token = struct {
    type: TokenType,
    pos: Pos,

    pub const Pos = struct {
        start: usize,
        end: usize,
    };
    pub const TokenType = enum {
        integer,
        float,
        rational,
        string,
        boolean,
        identifier,
        equals_equals,
        bang_equals,
        less_than,
        greater_than,
        less_equals,
        greater_equals,
        equals,
        star,
        slash,
        plus,
        minus,
        bang,
        semicolon,
        colon,
        comma,
        dot,
        left_bracket,
        right_bracket,
        left_curly,
        right_curly,
        left_square,
        right_square,
        kw_if,
        kw_else,
        kw_while,
        kw_return,
        kw_var,
        kw_func,
        kw_struct,
        kw_interface,
        kw_impl,

        pub fn lexeme(tok_type: TokenType) ?[]const u8 {
            return switch (tok_type) {
                .integer, .float, .rational, .string, .boolean, .identifier => null,
                .equals_equals => "==",
                .bang_equals => "!=",
                .less_than => "<",
                .greater_than => ">",
                .less_equals => "<=",
                .greater_equals => ">=",
                .equals => "=",
                .star => "*",
                .slash => "/",
                .plus => "+",
                .minus => "-",
                .bang => "!",
                .semicolon => ";",
                .colon => ":",
                .comma => ",",
                .dot => ".",
                .left_bracket => "(",
                .right_bracket => ")",
                .left_curly => "{",
                .right_curly => "}",
                .left_square => "[",
                .right_square => "]",
                .kw_if => "if",
                .kw_else => "else",
                .kw_while => "while",
                .kw_return => "return",
                .kw_var => "var",
                .kw_func => "fn",
                .kw_struct => "struct",
                .kw_interface => "interface",
                .kw_impl => "impl",
            };
        }
    };
    pub const keywords = std.StaticStringMap(TokenType).initComptime(.{
        .{ "true", .boolean },
        .{ "false", .boolean },
        .{ "if", .kw_if },
        .{ "else", .kw_else },
        .{ "while", .kw_while },
        .{ "return", .kw_return },
        .{ "var", .kw_var },
        .{ "fn", .kw_func },
        .{ "struct", .kw_struct },
        .{ "interface", .kw_interface },
        .{ "impl", .kw_impl },
    });
    pub const digraphs = std.StaticStringMap(TokenType).initComptime(.{
        .{ "==", .equals_equals },
        .{ "!=", .bang_equals },
        .{ "<=", .less_equals },
        .{ ">=", .greater_equals },
    });
    pub const monographs = std.StaticStringMap(TokenType).initComptime(.{
        .{ "<", .less_than },
        .{ ">", .greater_than },
        .{ "=", .equals },
        .{ "*", .star },
        .{ "/", .slash },
        .{ "+", .plus },
        .{ "-", .minus },
        .{ "!", .bang },
        .{ ";", .semicolon },
        .{ ":", .colon },
        .{ ",", .comma },
        .{ ".", .dot },
        .{ "(", .left_bracket },
        .{ ")", .right_bracket },
        .{ "{", .left_curly },
        .{ "}", .right_curly },
        .{ "[", .left_square },
        .{ "]", .right_square },
    });
    pub fn semicolon_insert(tok: Token) bool {
        return switch (tok.type) {
            .integer,
            .float,
            .rational,
            .string,
            .boolean,
            .identifier,
            .right_bracket,
            .right_square,
            .kw_return,
            => true,
            else => false,
        };
    }
};

fn tokenise(allocator: std.mem.Allocator, src: []const u8) []Token {
    var cur: usize = 0;
    var tokens = std.ArrayList(Token).init(allocator);
    while (cur < src.len) {
        var end: usize = 1;
        var typ: Token.TokenType = undefined;
        switch (src[cur]) {
            '\n' => {
                if (tokens.items.len > 0 and tokens.items[tokens.items.len - 2].semicolon_insert()) {
                    typ = .semicolon;
                    end = 0;
                } else {
                    cur += 1;
                    continue;
                }
            },
            ' ', '\t', '\r' => {
                cur += 1;
                continue;
            },
            'a'...'z', 'A'...'Z', '_' => {
                var i: usize = 0;
                while (cur + i < src.len) {
                    switch (src[cur + i]) {
                        'a'...'z', 'A'...'Z', '_' => i += 1,
                        else => {
                            break;
                        },
                    }
                }
                //print("{d}\n", .{}
                typ = Token.keywords.get(src[cur .. cur + i]) orelse .identifier;
                end = i;
            },
            '0'...'9' => {
                var i: usize = 0;
                typ = .integer;
                while (cur + i < src.len) {
                    i += 1;
                    if (typ == .integer) {
                        switch (src[cur + i]) {
                            '.' => {
                                if (std.ascii.isDigit(src[cur + i + 1])) {
                                    typ = .float;
                                    i += 1;
                                }
                            },
                            '/' => {
                                if (src[cur + i + 1] == '/' and std.ascii.isDigit(src[cur + i + 2])) {
                                    i += 2;
                                    typ = .rational;
                                }
                            },
                            else => {},
                        }
                    }
                    if (!std.ascii.isDigit(src[cur + i])) {
                        break;
                    }
                }
                end = i;
            },
            '"' => {
                var i: usize = 1;
                while (src[cur + i] != '"') {
                    i += 1;
                    if (cur + i > src.len) {
                        break;
                    }
                }
                i += 1;
                typ = .string;
                end = i;
            },
            else => {
                const mg = Token.monographs.get(src[cur .. cur + 1]);
                if (mg != null) {
                    typ = mg.?;
                    end = 1;
                }
                if (cur + 1 < src.len) {
                    const dg = Token.digraphs.get(src[cur .. cur + 2]);
                    if (dg != null) {
                        typ = dg.?;
                        end = 2;
                    }
                }
            },
        }
        const pos = Token.Pos{ .start = cur, .end = cur + end };
        cur += end;
        const tok = Token{ .type = typ, .pos = pos };
        tokens.append(tok) catch {
            continue;
        };
    }
    return tokens.items;
}

pub fn main() void {
    const src = "test 123 \"string\";";
    const allocator = std.heap.page_allocator;
    const toks = tokenise(allocator, src);
    for (toks) |tok| {
        print("tok: {s}\n", .{src[tok.pos.start..tok.pos.end]});
    }
}
